### DISCUSSION TITLE
Fail to set the DIMENSION entities' layer

### DISCUSSION THREAD

**Discussion by SeanXunIC** [mozman/ezdxf#1152] [Q&A] [answered]
Thanks for your attention! 
I tried to used the ezdxf **entity.set_dxf_attrib** function to change the layer of dimension from **_original_DIM_layer_** to **_new_DIM_layer_** (please kindly refer to the images and the attached dxf file). However, although I found the Layer of the dimensions are shown as the **_new_DIM_layer_**, when I turned off the  **_new_DIM_layer_**, the dimensions are still shown. If I turn off the old one (**_original_DIM_layer_**), the dimensions are hidden. I'm wondering if there's something I'm doing wrong that's preventing me from successfully changing layers? 

I've attached my code and the dxf file, thanks again!

```
import ezdxf

dxf_ori_path = ".\\dxf_original.dxf"
dxf_save_path = ".\\dxf_saved.dxf"

doc = ezdxf.readfile(dxf_ori_path)
msp = doc.modelspace()

# set the dim layer to copy
ori_dim_layer_name = 'original_DIM_layer'
new_dim_layer_name = 'new_DIM_layer'

# copy dimensions to new layer and remove the old ones
for entity in list(msp.query('*[layer=="{}"]'.format(ori_dim_layer_name))): 
    if entity.dxftype() != 'DIMENSION':
        continue

    new_entity = entity.copy()
    new_entity.set_dxf_attrib('layer', new_dim_layer_name)

    msp.unlink_entity(entity)
    msp.add_entity(new_entity)

doc.saveas(dxf_save_path)

# Then open the saved dxf file in AutoCAD
```
![image](https://github.com/user-attachments/assets/02e6c6a3-2806-47c9-9bef-9f8b3cf31b50)
![image](https://github.com/user-attachments/assets/39a83760-9247-4c03-b2cc-b2abbd35bc60)

These are the dxf files on Google Drive
https://drive.google.com/drive/folders/1gRtUEGqntc-QcIA1e1f_SvfH-y8IrQ16?usp=drive_link

**Comment by mozman** [comment, marked as answer]
The DIMENSION entity stores the parameters for the dimension line. The visual representation is rendered into an anonymous geometry block "*Dnnnn" .

To change  the layer of the visual representation you have to re-render the dimension line - the result will look different than the rendering from CAD applications, or you change the layer of all entities in the  geometry block:

The copy/remove procedure of your code is not necessary, layers are not containers they are just attributes.

```Python
for dimension in msp.query(f'DIMENSION[layer=="{ori_dim_layer_name}"]'): 
    dimension.dxf.layer = new_dim_layer_name
    block = dimension.get_geometry_block()
    if block:
        for entity in block:
            entity.dxf.layer = new_dim_layer_name
```

**Reply by SeanXunIC** [reply]
Wow! Thank you so much, that's helpful!
